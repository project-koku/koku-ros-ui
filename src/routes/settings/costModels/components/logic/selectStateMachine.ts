import type { MachineConfig } from 'xstate';
import { Machine } from 'xstate';

export interface SelectMachineContext {
  selection?: string[];
}

export interface SelectMachineStates {
  states: {
    collapsed: any;
    expanded: any;
  };
}

export type SelectMachineEvents = { type: 'TOGGLE'; selection?: string } | { type: 'SELECT'; selection: string };

export const selectMachineConfig: MachineConfig<SelectMachineContext, SelectMachineStates, SelectMachineEvents> = {
  context: {
    selection: [],
  },
  initial: 'collapsed',
  states: {
    collapsed: {
      on: {
        TOGGLE: 'expanded',
      },
    },
    expanded: {
      on: {
        TOGGLE: 'collapsed',
        SELECT: {
          target: 'collapsed',
          actions: ['assignSelection'],
        },
      },
    },
  },
};

export const selectMachineState = Machine(selectMachineConfig);
