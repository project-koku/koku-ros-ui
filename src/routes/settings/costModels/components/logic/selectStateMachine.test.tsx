import { createModel } from '@xstate/test';
import type { Interpreter } from 'xstate';
import { assign, interpret } from 'xstate';

import type { SelectMachineContext, SelectMachineEvents, SelectMachineStates } from './selectStateMachine';
import { selectMachineState } from './selectStateMachine';

describe('selectmachine', () => {
  const newMachine = selectMachineState.withConfig({
    actions: {
      assignSelection: assign({
        selection: (_ctx, evt) => [evt.selection],
      }),
    },
  });

  const s = interpret(newMachine).onTransition(stt => {
    expect([stt.context, stt.toStrings()]).toMatchSnapshot();
  }) as Interpreter<SelectMachineContext, SelectMachineStates, SelectMachineEvents>;
  s.start();

  const tModel = createModel(newMachine).withEvents({
    TOGGLE: {
      exec: () => {
        s.send({ type: 'TOGGLE' });
      },
    },
    SELECT: {
      exec: () => {
        s.send({ type: 'SELECT', selection: '23' });
      },
    },
  });

  tModel.getShortestPathPlansTo('expanded').forEach(plan => {
    plan.paths.forEach(path => {
      it(`${path.description}: ${plan.description}`, () => {
        path.test({});
      });
    });
  });
});
