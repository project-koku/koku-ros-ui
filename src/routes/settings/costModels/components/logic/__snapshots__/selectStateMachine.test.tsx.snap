// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
[
  {
    "selection": [],
  },
  [
    "collapsed",
  ],
]
`;

exports[`selectmachine via TOGGLE: reaches state: "#(machine).expanded" ({"selection":[]}) 1`] = `
[
  {
    "selection": [],
  },
  [
    "expanded",
  ],
]
`;
